go/vt/wrangler/shard.go

```diff
@@ -51,6 +53,11 @@ func (wr *Wrangler) updateShardMaster(ctx context.Context, si *topo.ShardInfo, t
 				return fmt.Errorf("creating this tablet would override old master %v in shard %v/%v", topoproto.TabletAliasString(s.MasterAlias), si.Keyspace(), si.ShardName())
 			}
 			s.MasterAlias = tabletAlias
+			// We need to set the MasterTermStartTime whenever we update MasterAlias
+			// It is possible that this tablet does not have a MasterTermStartTime set
+			// on it, but all that means is that if another "MASTER" tablet comes along
+			// that does have MasterTermStartTime set, it will override.
+			s.MasterTermStartTime = logutil.TimeToProto(time.Now())
```

Comment:
I'm having second thoughts now that I see this. It seems weird that this time is not coming from the tablet. Also, didn't we want one of the rules to be that the tablet itself is the one that updates MasterAlias?

Can you remind me why we thought it's important to set the timestamp here? Would it not be fine (and more in line with our rules) if we wait and let the tablet set it when it starts up?

Reply:
This gives rise to a corner case we did not cover: What is the timestamp of a master vttablet that was restarted? We should probably discuss this in a meeting.

Reply:
Currently, a master that's restarted will use `time.Now()` (when it comes back up) as its new timestamp, as long as it's still master according to the shard record:

https://github.com/vitessio/vitess/blob/74ca3b19cd1c10b50ff99614a08a22135477a35c/go/vt/vttablet/tabletmanager/init_tablet.go#L101-L125

With the new background shard sync we do, that tablet will also update the timestamp in the shard record. That effectively means "tablet restarted" is treated as an authoritative decision of who the master is, which should not be the case. A new master may have been elected but has not yet updated the shard record. If the old master restarts before the shard record is updated, its timestamp will win.

To fix this, perhaps we can say, if a tablet restarts and is still master in the shard record, it should read the timestamp from the shard record and set that as its internal timestamp.

Reply:
> To fix this, perhaps we can say, if a tablet restarts and is still master in the shard record, it should read the timestamp from the shard record and set that as its internal timestamp.

This makes sense.

Going back to the vtctl `InitTablet` command, when I look at the integration tests, it is possible to call `InitTablet` before a tablet is started for the first time.
That flow seems hokey to me though - the topology says a certain tablet is master, but it doesn't exist yet.
Maybe what we _should_ be doing is not update the shard record in InitTablet at all. If the tablet comes up as MASTER, it should update the shard record.

Reply:
It is worth noting that `InitTablet` is considered a low-level command that is allowed to change the topo information.